doc/tapscript_opcodes.md

```diff
@@ -0,0 +1,134 @@
+
+1. **Streaming Opcodes for streaming hashes**: There is an existing MAX_SCRIPT_ELEMENT_SIZE is 520 bytes, that is the maximum message size that can OP_SHA256 can operate on. This allows hashing on more than 520 bytes while still preserving the existing security against resource exhaustion attacks. The proposal for this is still under discussion in https://github.com/ElementsProject/elements/pull/817. We suggest the latest scheme suggested by Russel O Connor
+   -  Define OP_SUCCESS200 as a multi-byte opcode OP_STREAMINGHASH with the following semantics. Execute the push opcode following OP_STREAMINGHASH, if the opcode is not a push opcode, fail.
+   -  Pop the stack pop as minimal CScriptNum as n.
+      1. If n=0, interpret `OP_SHA256INITIALIZE` to pop a bytestring and push SHA256 context creating by adding the bytestring to the initial SHA256 context.
+      2. If n=1, interpret as `OP_SHA256UPDATE` to pop a SHA256 context and bytestring and push an updated context by adding the bytestring to the data stream being hashed.
+      3. If n=2, interpret as `OP_SHA256FINALIZE` to pop a SHA256 context and bytestring and push a SHA256 hash value after adding the bytestring and completing the padding.
+      4. If n=3, interpret `OP_RIPEMD160INITIALIZE` to pop a bytestring and push RIPEMD160 context creating by adding the bytestring to the initial RIPEMD160 context.
+      5. If n=4, interpret as `OP_RIPEMD160UPDATE` to pop a RIPEMD160 context and bytestring and push an updated context by adding the bytestring to the data stream being hashed.
+      6. If n=5, interpret as `OP_RIPEMD160FINALIZE` to pop a RIPEMD160 context and bytestring and push a RIPEMD160 hash value after adding the bytestring and completing the padding.
+      7. If n=6, interpret `OP_HASH256INITIALIZE` to pop a bytestring and push HASH256 context creating by adding the bytestring to the initial HASH256 context.
+      8. If n=7, interpret as `OP_HASH256UPDATE` to pop a HASH256 context and bytestring and push an updated context by adding the bytestring to the data stream being hashed.
+      9. If n=8, interpret as `OP_HASH256FINALIZE` to pop a HASH256 context and bytestring and push a HASH256 hash value after adding the bytestring and completing the padding.
+      10. If n=9, interpret `OP_HASH160INITIALIZE` to pop a bytestring and push HASH160 context creating by adding the bytestring to the initial HASH160 context.
+      11. If n=10, interpret as `OP_HASH160UPDATE` to pop a HASH160 context and bytestring and push an updated context by adding the bytestring to the data stream being hashed.
+      12. If n=11, interpret as `OP_HASH160FINALIZE` to pop a HASH160 context and bytestring and push a HASH160 hash value after adding the bytestring and completing the padding.
```

Comment:
What is the context when hash algorithms besides SHA256 would be needed ?

Reply:
The goal is to be as expressive as possible without much additional cost. Since elements already has these implemented, adding support would not be hard. I don't know where this might be needed, but here are some guesses that I can make.
- sha256d for verifying bitcoin merkle-root(I need to double-check if bitcoin actually uses sha256d for this). 
- Usecases involving verification of some p2sh script hash.
- Users may want a 20 byte digest for saving fees instead of a 32 byte one.

Reply:
fair enough

Reply:
In favor for simplicity, I think we should go with your suggestion of only requiring sha256